File: solutions/go/merge-sort.go
package main

import (
  "fmt"
  "math"
)

func mergeSort(ary []int) []int {
  if len(ary) <= 1 {
    return ary
  }
  left := make([]int, 0)
  right := make([]int, 0)
  middle := math.Floor(float64(len(ary) / 2))

  left = ary[0:int(middle)]
  right = ary[int(middle):]

  return merge(mergeSort(left), mergeSort(right))
}

func merge(left, right []int) []int {

  result := make([]int, 0)

  for len(left) > 0 || len(right) > 0 {
    if len(left) > 0 && len(right) > 0 {
      if left[0] <= right[0] {
        result = append(result, left[0])
        left = left[1:len(left)]
      } else {
        result = append(result, right[0])
        right = right[1:len(right)]
      }
    } else if len(left) > 0 {
      result = append(result, left[0])
      left = left[1:len(left)]
    } else {
      result = append(result, right[0])
      right = right[1:len(right)]
    }
  }

  return result
}

func main() {
  random := []int{1,73,7,1,72,58,933,574,24,74,52}
  fmt.Println("Unsorted: ", random)
  random = mergeSort(random)
  fmt.Println("Sorted: ", random)
}

File: solutions/java/MergeSort.java
/*
 * #### Merge Sort ####
 * Time Complexity : O(n log(n))
 * Space Complexity: O(n)
 *
 * */
public class MergeSort {
    public static void mergesort(int[] array) {
        int[] helper = new int[array.length];
        mergesort(array, helper, 0, array.length - 1);
    }

    public static void mergesort(int[] array, int[] helper, int low, int high) {
        if (low < high) {
            int middle = (low + high) / 2;
            mergesort(array, helper, low, middle); // Sort left half
            mergesort(array, helper, middle+1, high); // Sort right half
            merge(array, helper, low, middle, high); // Merge them
        }
    }

    public static void merge(int[] array, int[] helper, int low, int middle, int high) {
        /* Copy both halves into a helper array */
        for (int i = low; i <= high; i++) {
            helper[i] = array[i];
        }

        int helperLeft = low;
        int helperRight = middle + 1;
        int current = low;

        /* Iterate through helper array. Compare the left and right
         * half, copying back the smaller element from the two halves
         * into the original array. */
        while (helperLeft <= middle && helperRight <= high) {
            if (helper[helperLeft] <= helper[helperRight]) {
                array[current] = helper[helperLeft];
                helperLeft++;
            } else { // If right element is smaller than left element
                array[current] = helper[helperRight];
                helperRight++;
            }
            current++;
        }

        /* Copy the rest of the left side of the array into the
         * target array */
        int remaining = middle - helperLeft;
        for (int i = 0; i <= remaining; i++) {
            array[current + i] = helper[helperLeft + i];
        }
    }

    public static void main(String[] args) {
        int[] arr = {5, 4, 3, 2, 1, 8, 6, 7, 9};
        mergesort(arr);
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
    }

}